# Data wrangling 

We have the information of Canadian immigration form 1980 - 2013. We would like to analize the pattern of immigration to Canada from different countries. 

For example, how is the patter of immigration from Japan to Canada during these years ? 


The first thing we'll do is import two key data analysis modules: *pandas* and *numpy*.


In [1]:
import numpy as np  # useful for many scientific computing in Python
import pandas as pd # primary data structure library

In [2]:
df = pd.read_excel('datasets/Canada.xlsx',
                   sheet_name='Canada by Citizenship',
                   skiprows=range(20),
                   skipfooter=2,
                   engine='openpyxl'
                  )

Let's view the top 5 rows of the dataset using the `head()` function.


In [3]:
df.head(2)

Type    Coverage       OdName  AREA AreaName   REG          RegName  \
0  Immigrants  Foreigners  Afghanistan   935     Asia  5501    Southern Asia   
1  Immigrants  Foreigners      Albania   908   Europe   925  Southern Europe   

   DEV             DevName  1980  ...  2004  2005  2006  2007  2008  2009  \
0  902  Developing regions    16  ...  2978  3436  3009  2652  2111  1746   
1  901   Developed regions     1  ...  1450  1223   856   702   560   716   

   2010  2011  2012  2013  
0  1758  2203  2635  2004  
1   561   539   620   603  

[2 rows x 43 columns]

We can also view the bottom 5 rows of the dataset using the `tail()` function.


In [4]:
df.tail()

Type    Coverage          OdName  AREA AreaName  REG  \
190  Immigrants  Foreigners        Viet Nam   935     Asia  920   
191  Immigrants  Foreigners  Western Sahara   903   Africa  912   
192  Immigrants  Foreigners           Yemen   935     Asia  922   
193  Immigrants  Foreigners          Zambia   903   Africa  910   
194  Immigrants  Foreigners        Zimbabwe   903   Africa  910   

                RegName  DEV             DevName  1980  ...  2004  2005  2006  \
190  South-Eastern Asia  902  Developing regions  1191  ...  1816  1852  3153   
191     Northern Africa  902  Developing regions     0  ...     0     0     1   
192        Western Asia  902  Developing regions     1  ...   124   161   140   
193      Eastern Africa  902  Developing regions    11  ...    56    91    77   
194      Eastern Africa  902  Developing regions    72  ...  1450   615   454   

     2007  2008  2009  2010  2011  2012  2013  
190  2574  1784  2171  1942  1723  1731  2112  
191     0     0     0     0     0     0     0  
192   122   133   128   211   160   174   217  
193    71    64    60   102    69    46    59  
194   663   611   508   494   434   437   407  

[5 rows x 43 columns]

Note that, here we could not see all the columns. To see the entire columns and rows we have to specify the number of rows and columns we want to see. 

**Setting max columns and rows**
```Python
pd.set_option("display.max_columns", None, #Allows all columns
              "display.max_rows", 500      #Allows 500 rows 
             )

```

In [5]:
pd.set_option("display.max_columns", None)

In [6]:
df.tail(2)

Type    Coverage    OdName  AREA AreaName  REG         RegName  \
193  Immigrants  Foreigners    Zambia   903   Africa  910  Eastern Africa   
194  Immigrants  Foreigners  Zimbabwe   903   Africa  910  Eastern Africa   

     DEV             DevName  1980  1981  1982  1983  1984  1985  1986  1987  \
193  902  Developing regions    11    17    11     7    16     9    15    23   
194  902  Developing regions    72   114   102    44    32    29    43    68   

     1988  1989  1990  1991  1992  1993  1994  1995  1996  1997  1998  1999  \
193    44    68    77    69    73    46    51    41    34    72    34    51   
194    99   187   129    94    61    72    78    58    39    44    43    49   

     2000  2001  2002  2003  2004  2005  2006  2007  2008  2009  2010  2011  \
193    39    78    50    46    56    91    77    71    64    60   102    69   
194    98   110   191   669  1450   615   454   663   611   508   494   434   

     2012  2013  
193    46    59  
194   437   407

In [7]:
df

Type    Coverage          OdName  AREA AreaName   REG  \
0    Immigrants  Foreigners     Afghanistan   935     Asia  5501   
1    Immigrants  Foreigners         Albania   908   Europe   925   
2    Immigrants  Foreigners         Algeria   903   Africa   912   
3    Immigrants  Foreigners  American Samoa   909  Oceania   957   
4    Immigrants  Foreigners         Andorra   908   Europe   925   
..          ...         ...             ...   ...      ...   ...   
190  Immigrants  Foreigners        Viet Nam   935     Asia   920   
191  Immigrants  Foreigners  Western Sahara   903   Africa   912   
192  Immigrants  Foreigners           Yemen   935     Asia   922   
193  Immigrants  Foreigners          Zambia   903   Africa   910   
194  Immigrants  Foreigners        Zimbabwe   903   Africa   910   

                RegName  DEV             DevName  1980  1981  1982  1983  \
0         Southern Asia  902  Developing regions    16    39    39    47   
1       Southern Europe  901   Developed regions     1     0     0     0   
2       Northern Africa  902  Developing regions    80    67    71    69   
3             Polynesia  902  Developing regions     0     1     0     0   
4       Southern Europe  901   Developed regions     0     0     0     0   
..                  ...  ...                 ...   ...   ...   ...   ...   
190  South-Eastern Asia  902  Developing regions  1191  1829  2162  3404   
191     Northern Africa  902  Developing regions     0     0     0     0   
192        Western Asia  902  Developing regions     1     2     1     6   
193      Eastern Africa  902  Developing regions    11    17    11     7   
194      Eastern Africa  902  Developing regions    72   114   102    44   

     1984  1985  1986  1987  1988  1989  1990  1991  1992  1993  1994  1995  \
0      71   340   496   741   828  1076  1028  1378  1170   713   858  1537   
1       0     0     1     2     2     3     3    21    56    96    71    63   
2      63    44    69   132   242   434   491   872   795   717   595  1106   
3       0     0     0     1     0     1     2     0     0     0     0     0   
4       0     0     2     0     0     0     3     0     1     0     0     0   
..    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
190  7583  5907  2741  1406  1411  3004  3801  5870  5416  6547  5105  3723   
191     0     0     0     0     0     0     0     1     0     0     0     0   
192     0    18     7    12     7    18     4    18    41    41    39    73   
193    16     9    15    23    44    68    77    69    73    46    51    41   
194    32    29    43    68    99   187   129    94    61    72    78    58   

     1996  1997  1998  1999  2000  2001  2002  2003  2004  2005  2006  2007  \
0    2212  2555  1999  2395  3326  4067  3697  3479  2978  3436  3009  2652   
1     113   307   574  1264  1816  1602  1021   853  1450  1223   856   702   
2    2054  1842  2292  2389  2867  3418  3406  3072  3616  3626  4807  3623   
3       0     0     0     0     0     0     0     0     0     0     1     0   
4       0     0     2     0     0     1     0     2     0     0     1     1   
..    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
190  2462  1752  1631  1419  1803  2117  2291  1713  1816  1852  3153  2574   
191     0     0     0     0     0     0     0     0     0     0     1     0   
192   144   121   141   134   122   181   171   113   124   161   140   122   
193    34    72    34    51    39    78    50    46    56    91    77    71   
194    39    44    43    49    98   110   191   669  1450   615   454   663   

     2008  2009  2010  2011  2012  2013  
0    2111  1746  1758  2203  2635  2004  
1     560   716   561   539   620   603  
2    4005  5393  4752  4325  3774  4331  
3       0     0     0     0     0     0  
4       0     0     0     0     1     1  
..    ...   ...   ...   ...   ...   ...  
190  1784  2171  1942  1723  1731  2112  
191     0     0     0     0     0     0  
192  

To start by getting basic information about your `dataframe`: 

In [8]:
df.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Columns: 43 entries, Type to 2013
dtypes: int64(37), object(6)
memory usage: 65.6+ KB


Size of dataframe is : (#rows, #columns)

In [9]:
df.shape  

(195, 43)

In [10]:
df.columns

Index([    'Type', 'Coverage',   'OdName',     'AREA', 'AreaName',      'REG',
        'RegName',      'DEV',  'DevName',       1980,       1981,       1982,
             1983,       1984,       1985,       1986,       1987,       1988,
             1989,       1990,       1991,       1992,       1993,       1994,
             1995,       1996,       1997,       1998,       1999,       2000,
             2001,       2002,       2003,       2004,       2005,       2006,
             2007,       2008,       2009,       2010,       2011,       2012,
             2013],
      dtype='object')

In [11]:
df.index

RangeIndex(start=0, stop=195, step=1)

Note: The default type of intance variables `index` and `columns` are **NOT** `list`.


In [12]:
print(type(df.columns))
print(type(df.index))

<class 'pandas.core.indexes.base.Index'>
<class 'pandas.core.indexes.range.RangeIndex'>


To get the index and columns as lists, we can use the `tolist()` method.


In [13]:
df.columns.tolist()

['Type',
 'Coverage',
 'OdName',
 'AREA',
 'AreaName',
 'REG',
 'RegName',
 'DEV',
 'DevName',
 1980,
 1981,
 1982,
 1983,
 1984,
 1985,
 1986,
 1987,
 1988,
 1989,
 1990,
 1991,
 1992,
 1993,
 1994,
 1995,
 1996,
 1997,
 1998,
 1999,
 2000,
 2001,
 2002,
 2003,
 2004,
 2005,
 2006,
 2007,
 2008,
 2009,
 2010,
 2011,
 2012,
 2013]

In [14]:
df.index.tolist()

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,


In [15]:
print(type(df.columns.tolist()))
print(type(df.index.tolist()))

<class 'list'>
<class 'list'>


To view the dimensions of the dataframe, we use the `shape` instance variable of it.


**Note**: The main types stored in *pandas* objects are `float`, `int`, `bool`, `datetime64[ns]`, `datetime64[ns, tz]`, `timedelta[ns]`, `category`, and `object` (string). In addition, these dtypes have item sizes, e.g. `int64` and `int32`.


In pandas 

* axis = 0  ; The row (default) and 

* axis = 1  ; The columns

Let's clean the data set to remove a few unnecessary columns. We can use *pandas* `drop()` method as follows:


In [16]:
df.head(2)

Type    Coverage       OdName  AREA AreaName   REG          RegName  \
0  Immigrants  Foreigners  Afghanistan   935     Asia  5501    Southern Asia   
1  Immigrants  Foreigners      Albania   908   Europe   925  Southern Europe   

   DEV             DevName  1980  1981  1982  1983  1984  1985  1986  1987  \
0  902  Developing regions    16    39    39    47    71   340   496   741   
1  901   Developed regions     1     0     0     0     0     0     1     2   

   1988  1989  1990  1991  1992  1993  1994  1995  1996  1997  1998  1999  \
0   828  1076  1028  1378  1170   713   858  1537  2212  2555  1999  2395   
1     2     3     3    21    56    96    71    63   113   307   574  1264   

   2000  2001  2002  2003  2004  2005  2006  2007  2008  2009  2010  2011  \
0  3326  4067  3697  3479  2978  3436  3009  2652  2111  1746  1758  2203   
1  1816  1602  1021   853  1450  1223   856   702   560   716   561   539   

   2012  2013  
0  2635  2004  
1   620   603

`Drop` the columns that we are not interested, by: 

In [17]:
favs = ['Type','Coverage','AREA','REG','DEV', 'DevName']

df.drop(favs, 
        axis=1, #axis One is Column
        inplace=True
       )

df.head(2)

OdName AreaName          RegName  1980  1981  1982  1983  1984  1985  \
0  Afghanistan     Asia    Southern Asia    16    39    39    47    71   340   
1      Albania   Europe  Southern Europe     1     0     0     0     0     0   

   1986  1987  1988  1989  1990  1991  1992  1993  1994  1995  1996  1997  \
0   496   741   828  1076  1028  1378  1170   713   858  1537  2212  2555   
1     1     2     2     3     3    21    56    96    71    63   113   307   

   1998  1999  2000  2001  2002  2003  2004  2005  2006  2007  2008  2009  \
0  1999  2395  3326  4067  3697  3479  2978  3436  3009  2652  2111  1746   
1   574  1264  1816  1602  1021   853  1450  1223   856   702   560   716   

   2010  2011  2012  2013  
0  1758  2203  2635  2004  
1   561   539   620   603

Let's `rename` the columns so that they make sense. 

We can use `rename()` method by passing in a `dictionary` of old and new names as follows:

In [18]:
new_names = {'OdName':'Country', 
            'AreaName':'Continent', 
            'RegName':'Region'
            }

type(new_names)

dict

In [19]:
df.rename(columns= new_names, 
          inplace=True       #replace the orginal df
         )
df.head(2)

Country Continent           Region  1980  1981  1982  1983  1984  1985  \
0  Afghanistan      Asia    Southern Asia    16    39    39    47    71   340   
1      Albania    Europe  Southern Europe     1     0     0     0     0     0   

   1986  1987  1988  1989  1990  1991  1992  1993  1994  1995  1996  1997  \
0   496   741   828  1076  1028  1378  1170   713   858  1537  2212  2555   
1     1     2     2     3     3    21    56    96    71    63   113   307   

   1998  1999  2000  2001  2002  2003  2004  2005  2006  2007  2008  2009  \
0  1999  2395  3326  4067  3697  3479  2978  3436  3009  2652  2111  1746   
1   574  1264  1816  1602  1021   853  1450  1223   856   702   560   716   

   2010  2011  2012  2013  
0  1758  2203  2635  2004  
1   561   539   620   603

We will also add a 'Total' column that sums up the total immigrants by country over the entire period 1980 - 2013, as follows:


In [20]:
df['Total'] = df.iloc[:, 4:].sum(axis=1) #axis=0 is row; axis=1 is column. Here sum of 'row'
df

Country Continent              Region  1980  1981  1982  1983  \
0       Afghanistan      Asia       Southern Asia    16    39    39    47   
1           Albania    Europe     Southern Europe     1     0     0     0   
2           Algeria    Africa     Northern Africa    80    67    71    69   
3    American Samoa   Oceania           Polynesia     0     1     0     0   
4           Andorra    Europe     Southern Europe     0     0     0     0   
..              ...       ...                 ...   ...   ...   ...   ...   
190        Viet Nam      Asia  South-Eastern Asia  1191  1829  2162  3404   
191  Western Sahara    Africa     Northern Africa     0     0     0     0   
192           Yemen      Asia        Western Asia     1     2     1     6   
193          Zambia    Africa      Eastern Africa    11    17    11     7   
194        Zimbabwe    Africa      Eastern Africa    72   114   102    44   

     1984  1985  1986  1987  1988  1989  1990  1991  1992  1993  1994  1995  \
0      71   340   496   741   828  1076  1028  1378  1170   713   858  1537   
1       0     0     1     2     2     3     3    21    56    96    71    63   
2      63    44    69   132   242   434   491   872   795   717   595  1106   
3       0     0     0     1     0     1     2     0     0     0     0     0   
4       0     0     2     0     0     0     3     0     1     0     0     0   
..    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
190  7583  5907  2741  1406  1411  3004  3801  5870  5416  6547  5105  3723   
191     0     0     0     0     0     0     0     1     0     0     0     0   
192     0    18     7    12     7    18     4    18    41    41    39    73   
193    16     9    15    23    44    68    77    69    73    46    51    41   
194    32    29    43    68    99   187   129    94    61    72    78    58   

     1996  1997  1998  1999  2000  2001  2002  2003  2004  2005  2006  2007  \
0    2212  2555  1999  2395  3326  4067  3697  3479  2978  3436  3009  2652   
1     113   307   574  1264  1816  1602  1021   853  1450  1223   856   702   
2    2054  1842  2292  2389  2867  3418  3406  3072  3616  3626  4807  3623   
3       0     0     0     0     0     0     0     0     0     0     1     0   
4       0     0     2     0     0     1     0     2     0     0     1     1   
..    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
190  2462  1752  1631  1419  1803  2117  2291  1713  1816  1852  3153  2574   
191     0     0     0     0     0     0     0     0     0     0     1     0   
192   144   121   141   134   122   181   171   113   124   161   140   122   
193    34    72    34    51    39    78    50    46    56    91    77    71   
194    39    44    43    49    98   110   191   669  1450   615   454   663   

     2008  2009  2010  2011  2012  2013  Total  
0    2111  1746  1758  2203  2635  2004  58623  
1     560   716   561   539   620   603  15698  
2    4005  5393  4752  4325  3774  4331  69359  
3       0     0     0     0     0     0      6  
4       0     0     0     0     1     1     15  
..    ...   ...   ...   ...   ...   ...    ...  
190  1784  2171  1942  1723  1731  2112  95955  
191     0     0     0     0     0     0      2  
192   133   128   211   160   174   217   2984  
193    64    60   102    69    46    59   1666  
194   611   508   494   434   437   407   8526  

[195 rows x 38 columns]

We can check to see how many null objects we have in the dataset as follows:


In [21]:
df.isnull().sum()

Country      0
Continent    0
Region       0
1980         0
1981         0
1982         0
1983         0
1984         0
1985         0
1986         0
1987         0
1988         0
1989         0
1990         0
1991         0
1992         0
1993         0
1994         0
1995         0
1996         0
1997         0
1998         0
1999         0
2000         0
2001         0
2002         0
2003         0
2004         0
2005         0
2006         0
2007         0
2008         0
2009         0
2010         0
2011         0
2012         0
2013         0
Total        0
dtype: int64

Finally, let's view a quick summary of each column in our dataframe using the `describe()` method.


## Descriptive statistics

In [22]:
df.describe()

1980          1981          1982          1983          1984  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean     508.394872    566.989744    534.723077    387.435897    376.497436   
std     1949.588546   2152.643752   1866.997511   1204.333597   1198.246371   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%       13.000000     10.000000     11.000000     12.000000     13.000000   
75%      251.500000    295.500000    275.000000    173.000000    181.000000   
max    22045.000000  24796.000000  20620.000000  10015.000000  10170.000000   

              1985         1986          1987          1988          1989  \
count   195.000000   195.000000    195.000000    195.000000    195.000000   
mean    358.861538   441.271795    691.133333    714.389744    843.241026   
std    1079.309600  1225.576630   2109.205607   2443.606788   2555.048874   
min       0.000000     0.000000      0.000000      0.000000      0.000000   
25%       0.000000     0.500000      0.500000      1.000000      1.000000   
50%      17.000000    18.000000     26.000000     34.000000     44.000000   
75%     197.000000   254.000000    434.000000    409.000000    508.500000   
max    9564.000000  9470.000000  21337.000000  27359.000000  23795.000000   

               1990          1991          1992          1993          1994  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean     964.379487   1064.148718   1136.856410   1138.712821    993.153846   
std     3158.730195   2952.093731   3330.083742   3495.220063   3613.336444   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        2.000000      3.000000      6.500000     11.500000      9.500000   
50%       38.000000     51.000000     74.000000     85.000000     76.000000   
75%      612.500000    657.500000    655.000000    722.500000    545.000000   
max    31668.000000  23380.000000  34123.000000  33720.000000  39231.000000   

               1995          1996          1997          1998          1999  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean     962.625641   1026.076923    989.153846    824.241026    922.143590   
std     3091.492343   3321.045004   3070.761447   2385.943695   2887.632585   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       10.500000     14.500000     19.500000     15.000000     16.000000   
50%       91.000000    118.000000    114.000000    106.000000    116.000000   
75%      550.500000    603.500000    612.500000    535.500000    548.500000   
max    30145.000000  29322.000000  22965.000000  21049.000000  30069.000000   

               2000          2001          2002          2003          2004  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean    1111.343590   1244.323077   1144.158974   1114.343590   1190.169231   
std     3664.042361   3961.621410   3660.579836   3623.509519   3710.505369   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       16.000000     22.000000     18.500000     21.500000     19.000000   
50%      138.000000    169.000000    165.000000    161.000000    191.000000   
75%      659.000000    793.500000    686.000000    673.500000    756.500000   
max    35529.000000  36434.000000  31961.000000  36439.000000  36619.000000   

               2005          2006          2007          2008          2009  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean    1320.292308   1266.958974   1191.820513   1246.394872   1275.733333   
std     4425.957828   3926.717747   3443.542409   3694.573544   3829.630424   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       28.500000     25.000000     31.000000     31.000000     36.000000   

##   Indexing and Selection (slicing)


### Select Column

**There are two ways to filter on a column name:**

Method 1: Quick and easy, but only works if the column name does NOT have spaces or special characters.

```python
    df.columnName               # returns series
```

Method 2: More robust, and can filter on multiple columns.

```python
    df['column']                  # returns series
```

```python
    df[['column 1', 'column 2']]  # returns dataframe; see  '[[' & ']]' 
```

***


Example: Let's try filtering on the list of countries ('Country').


In [23]:
df.Country  # returns a series

0         Afghanistan
1             Albania
2             Algeria
3      American Samoa
4             Andorra
            ...      
190          Viet Nam
191    Western Sahara
192             Yemen
193            Zambia
194          Zimbabwe
Name: Country, Length: 195, dtype: object

In [24]:
type(df.Country)

pandas.core.series.Series

In [25]:
df['Country']

0         Afghanistan
1             Albania
2             Algeria
3      American Samoa
4             Andorra
            ...      
190          Viet Nam
191    Western Sahara
192             Yemen
193            Zambia
194          Zimbabwe
Name: Country, Length: 195, dtype: object

In [26]:
type(df['Country'])

pandas.core.series.Series

In [27]:
df[['Country']] #dataframe 

Country
0       Afghanistan
1           Albania
2           Algeria
3    American Samoa
4           Andorra
..              ...
190        Viet Nam
191  Western Sahara
192           Yemen
193          Zambia
194        Zimbabwe

[195 rows x 1 columns]

In [28]:
type(df[['Country']])

pandas.core.frame.DataFrame

Let's try filtering on the list of countries ('Country') and the data for years: 1980 - 1985.


In [29]:
df[['Country', 1980, 1981, 1982, 1983, 1984, 1985]] # returns a dataframe

Country  1980  1981  1982  1983  1984  1985
0       Afghanistan    16    39    39    47    71   340
1           Albania     1     0     0     0     0     0
2           Algeria    80    67    71    69    63    44
3    American Samoa     0     1     0     0     0     0
4           Andorra     0     0     0     0     0     0
..              ...   ...   ...   ...   ...   ...   ...
190        Viet Nam  1191  1829  2162  3404  7583  5907
191  Western Sahara     0     0     0     0     0     0
192           Yemen     1     2     1     6     0    18
193          Zambia    11    17    11     7    16     9
194        Zimbabwe    72   114   102    44    32    29

[195 rows x 7 columns]

### Select Row

There are main 2 ways to select rows:

```python
    df.loc[label]    # filters by the labels of the index/column
    df.iloc[index]   # filters by the positions of the index/column
```


In [30]:
df.loc[2].head() # The row with index =  2

Country              Algeria
Continent             Africa
Region       Northern Africa
1980                      80
1981                      67
Name: 2, dtype: object

In [31]:
df.head(2)

Country Continent           Region  1980  1981  1982  1983  1984  1985  \
0  Afghanistan      Asia    Southern Asia    16    39    39    47    71   340   
1      Albania    Europe  Southern Europe     1     0     0     0     0     0   

   1986  1987  1988  1989  1990  1991  1992  1993  1994  1995  1996  1997  \
0   496   741   828  1076  1028  1378  1170   713   858  1537  2212  2555   
1     1     2     2     3     3    21    56    96    71    63   113   307   

   1998  1999  2000  2001  2002  2003  2004  2005  2006  2007  2008  2009  \
0  1999  2395  3326  4067  3697  3479  2978  3436  3009  2652  2111  1746   
1   574  1264  1816  1602  1021   853  1450  1223   856   702   560   716   

   2010  2011  2012  2013  Total  
0  1758  2203  2635  2004  58623  
1   561   539   620   603  15698

Notice that the default index of the dataset is a numeric range from 0 to 194. 

This makes it very difficult to do a query by a specific country. For example to search for data on Japan, we need to know the corresponding index value.

This can be fixed very easily by setting the 'Country' column as the index using `set_index()` method.


In [32]:
df.set_index('Country', inplace=True) #inplace=True - repalcing the orginal dataset
df.head(2)

Continent           Region  1980  1981  1982  1983  1984  1985  \
Country                                                                      
Afghanistan      Asia    Southern Asia    16    39    39    47    71   340   
Albania        Europe  Southern Europe     1     0     0     0     0     0   

             1986  1987  1988  1989  1990  1991  1992  1993  1994  1995  1996  \
Country                                                                         
Afghanistan   496   741   828  1076  1028  1378  1170   713   858  1537  2212   
Albania         1     2     2     3     3    21    56    96    71    63   113   

             1997  1998  1999  2000  2001  2002  2003  2004  2005  2006  2007  \
Country                                                                         
Afghanistan  2555  1999  2395  3326  4067  3697  3479  2978  3436  3009  2652   
Albania       307   574  1264  1816  1602  1021   853  1450  1223   856   702   

             2008  2009  2010  2011  2012  2013  Total  
Country                                                 
Afghanistan  2111  1746  1758  2203  2635  2004  58623  
Albania       560   716   561   539   620   603  15698

Tip: The opposite of `set_index()` is `reset_index()`. 
So to reset the index, we can use df.reset_index()

In [33]:
df=df.reset_index()
df.head(3)

Country Continent           Region  1980  1981  1982  1983  1984  1985  \
0  Afghanistan      Asia    Southern Asia    16    39    39    47    71   340   
1      Albania    Europe  Southern Europe     1     0     0     0     0     0   
2      Algeria    Africa  Northern Africa    80    67    71    69    63    44   

   1986  1987  1988  1989  1990  1991  1992  1993  1994  1995  1996  1997  \
0   496   741   828  1076  1028  1378  1170   713   858  1537  2212  2555   
1     1     2     2     3     3    21    56    96    71    63   113   307   
2    69   132   242   434   491   872   795   717   595  1106  2054  1842   

   1998  1999  2000  2001  2002  2003  2004  2005  2006  2007  2008  2009  \
0  1999  2395  3326  4067  3697  3479  2978  3436  3009  2652  2111  1746   
1   574  1264  1816  1602  1021   853  1450  1223   856   702   560   716   
2  2292  2389  2867  3418  3406  3072  3616  3626  4807  3623  4005  5393   

   2010  2011  2012  2013  Total  
0  1758  2203  2635  2004  58623  
1   561   539   620   603  15698  
2  4752  4325  3774  4331  69359

In [34]:
df.set_index('Country', inplace=True)
df.head(3)

Continent           Region  1980  1981  1982  1983  1984  1985  \
Country                                                                      
Afghanistan      Asia    Southern Asia    16    39    39    47    71   340   
Albania        Europe  Southern Europe     1     0     0     0     0     0   
Algeria        Africa  Northern Africa    80    67    71    69    63    44   

             1986  1987  1988  1989  1990  1991  1992  1993  1994  1995  1996  \
Country                                                                         
Afghanistan   496   741   828  1076  1028  1378  1170   713   858  1537  2212   
Albania         1     2     2     3     3    21    56    96    71    63   113   
Algeria        69   132   242   434   491   872   795   717   595  1106  2054   

             1997  1998  1999  2000  2001  2002  2003  2004  2005  2006  2007  \
Country                                                                         
Afghanistan  2555  1999  2395  3326  4067  3697  3479  2978  3436  3009  2652   
Albania       307   574  1264  1816  1602  1021   853  1450  1223   856   702   
Algeria      1842  2292  2389  2867  3418  3406  3072  3616  3626  4807  3623   

             2008  2009  2010  2011  2012  2013  Total  
Country                                                 
Afghanistan  2111  1746  1758  2203  2635  2004  58623  
Albania       560   716   561   539   620   603  15698  
Algeria      4005  5393  4752  4325  3774  4331  69359

Optional: To remove the name of the `index`

In [35]:
df.index.name = None
df.head(2)

Continent           Region  1980  1981  1982  1983  1984  1985  \
Afghanistan      Asia    Southern Asia    16    39    39    47    71   340   
Albania        Europe  Southern Europe     1     0     0     0     0     0   

             1986  1987  1988  1989  1990  1991  1992  1993  1994  1995  1996  \
Afghanistan   496   741   828  1076  1028  1378  1170   713   858  1537  2212   
Albania         1     2     2     3     3    21    56    96    71    63   113   

             1997  1998  1999  2000  2001  2002  2003  2004  2005  2006  2007  \
Afghanistan  2555  1999  2395  3326  4067  3697  3479  2978  3436  3009  2652   
Albania       307   574  1264  1816  1602  1021   853  1450  1223   856   702   

             2008  2009  2010  2011  2012  2013  Total  
Afghanistan  2111  1746  1758  2203  2635  2004  58623  
Albania       560   716   561   539   620   603  15698

Example: Let's view the number of immigrants from Japan (row 87) for the following scenarios:

1. The full row data (all columns)
2. For year 2013
3. For years 1980 to 1985

**1. The full row data (all columns)**

In [36]:
df.loc['Japan'].head() #returns list

Continent            Asia
Region       Eastern Asia
1980                  701
1981                  756
1982                  598
Name: Japan, dtype: object

In [37]:
df.loc[['Japan']] #returns dataframe

Continent        Region  1980  1981  1982  1983  1984  1985  1986  1987  \
Japan      Asia  Eastern Asia   701   756   598   309   246   198   248   422   

       1988  1989  1990  1991  1992  1993  1994  1995  1996  1997  1998  1999  \
Japan   324   494   379   506   605   907   956   826   994   924   897  1083   

       2000  2001  2002  2003  2004  2005  2006  2007  2008  2009  2010  2011  \
Japan  1010  1092   806   817   973  1067  1212  1250  1284  1194  1168  1265   

       2012  2013  Total  
Japan  1214   982  27006

Alternate methods :

In [38]:
df.iloc[87].head()

Continent            Asia
Region       Eastern Asia
1980                  701
1981                  756
1982                  598
Name: Japan, dtype: object

In [39]:
type(df.iloc[87])

pandas.core.series.Series

In [40]:
df.iloc[[87]] #dataframe

Continent        Region  1980  1981  1982  1983  1984  1985  1986  1987  \
Japan      Asia  Eastern Asia   701   756   598   309   246   198   248   422   

       1988  1989  1990  1991  1992  1993  1994  1995  1996  1997  1998  1999  \
Japan   324   494   379   506   605   907   956   826   994   924   897  1083   

       2000  2001  2002  2003  2004  2005  2006  2007  2008  2009  2010  2011  \
Japan  1010  1092   806   817   973  1067  1212  1250  1284  1194  1168  1265   

       2012  2013  Total  
Japan  1214   982  27006

In [41]:
type(df.iloc[[87]])

pandas.core.frame.DataFrame

In [42]:
df[df.index == 'Japan']

Continent        Region  1980  1981  1982  1983  1984  1985  1986  1987  \
Japan      Asia  Eastern Asia   701   756   598   309   246   198   248   422   

       1988  1989  1990  1991  1992  1993  1994  1995  1996  1997  1998  1999  \
Japan   324   494   379   506   605   907   956   826   994   924   897  1083   

       2000  2001  2002  2003  2004  2005  2006  2007  2008  2009  2010  2011  \
Japan  1010  1092   806   817   973  1067  1212  1250  1284  1194  1168  1265   

       2012  2013  Total  
Japan  1214   982  27006

**2. For year 2013**

In [43]:
df.loc['Japan', 2013]

982

Alternative: 

year 2013 is the last column, with a positional index of Japan is 36

In [44]:
df.iloc[87, 35]

982

**3. for years 1980 to 1985**

In [45]:
fav_years = [1980, 1981, 1982, 1983, 1984, 1985]
df.loc['Japan',fav_years]

1980    701
1981    756
1982    598
1983    309
1984    246
1985    198
Name: Japan, dtype: object

Alternative:

In [46]:
df.iloc[87, [2, 3, 4, 5, 6, 7]]

1980    701
1981    756
1982    598
1983    309
1984    246
1985    198
Name: Japan, dtype: object

Column names that are integers (such as the years) might introduce some confusion. 

For example, when we are referencing the year 2013, one might confuse that when the 2013th positional index.

To avoid this ambuigity, let's convert the column names into strings: '1980' to '2013'.


In [47]:
df.columns = list(map(str, df.columns))

#[print (type(x)) for x in df.columns.values]

Since we converted the years to string, let's declare a variable that will allow us to easily call upon the full range of years:


In [48]:
years = list(map(str, range(1980, 2014)))
years[:3] #from the begining to the thrid item

['1980', '1981', '1982']

## Filtering based on a criteria

To filter the dataframe based on a condition, we simply pass the condition as a boolean vector.

For example, Let's filter the dataframe to show the data on Asian countries (AreaName = Asia).

In [49]:
condition = df['Continent'] == 'Asia'   # creating the condition `boolean` series
print(condition)

Afghanistan        True
Albania           False
Algeria           False
American Samoa    False
Andorra           False
                  ...  
Viet Nam           True
Western Sahara    False
Yemen              True
Zambia            False
Zimbabwe          False
Name: Continent, Length: 195, dtype: bool


In [50]:
df[condition]    # passing this condition into the DataFrame

Continent              Region  \
Afghanistan                                         Asia       Southern Asia   
Armenia                                             Asia        Western Asia   
Azerbaijan                                          Asia        Western Asia   
Bahrain                                             Asia        Western Asia   
Bangladesh                                          Asia       Southern Asia   
Bhutan                                              Asia       Southern Asia   
Brunei Darussalam                                   Asia  South-Eastern Asia   
Cambodia                                            Asia  South-Eastern Asia   
China                                               Asia        Eastern Asia   
China, Hong Kong Special Administrative Region      Asia        Eastern Asia   
China, Macao Special Administrative Region          Asia        Eastern Asia   
Cyprus                                              Asia        Western Asia   
Democratic People's Republic of Korea               Asia        Eastern Asia   
Georgia                                             Asia        Western Asia   
India                                               Asia       Southern Asia   
Indonesia                                           Asia  South-Eastern Asia   
Iran (Islamic Republic of)                          Asia       Southern Asia   
Iraq                                                Asia        Western Asia   
Israel                                              Asia        Western Asia   
Japan                                               Asia        Eastern Asia   
Jordan                                              Asia        Western Asia   
Kazakhstan                                          Asia        Central Asia   
Kuwait                                              Asia        Western Asia   
Kyrgyzstan                                          Asia        Central Asia   
Lao People's Democratic Republic                    Asia  South-Eastern Asia   
Lebanon                                             Asia        Western Asia   
Malaysia                                            Asia  South-Eastern Asia   
Maldives                                            Asia       Southern Asia   
Mongolia                                            Asia        Eastern Asia   
Myanmar                                             Asia  South-Eastern Asia   
Nepal                                               Asia       Southern Asia   
Oman                                                Asia        Western Asia   
Pakistan                                            Asia       Southern Asia   
Philippines                                         Asia  South-Eastern Asia   
Qatar                                               Asia        Western Asia   
Republic of Korea                                   Asia        Eastern Asia   
Saudi Arabia                                        Asia        Western Asia   
Singapore                                           Asia  South-Eastern Asia   
Sri Lanka                                           Asia       Southern Asia   
State of Palestine                                  Asia        Western Asia   
Syrian Arab Republic                                Asia        Western Asia   
Tajikistan                                          Asia        Central Asia   
Thailand                                            Asia  South-Eastern Asia   
Turkey                                              Asia        Western Asia   
Turkmenistan                                        Asia        Central Asia   
United Arab Emirates                                Asia        Western Asia   
Uzbekistan                                          Asia        Central Asia   
Viet Nam                                            Asia  South-Eastern Asia   
Yemen                                               Asia        Western Asia   

                                              

Which gives us the data only belongs to Asia. 

We can pass multiple criteria in the same line.
```Python
    Condition1 = AreaNAme == Asia 
    Condition2 = RegName == Southern Asia
```

In [51]:
Condition1 = df['Continent']=='Asia'
Condition2 = df['Region']=='Southern Asia'
Condition3 = df['Total'] < 6000

df[ Condition1 & Condition2 & Condition3]


Continent         Region  1980  1981  1982  1983  1984  1985  1986  \
Bhutan        Asia  Southern Asia     0     0     0     0     1     0     0   
Maldives      Asia  Southern Asia     0     0     0     1     0     0     0   

          1987  1988  1989  1990  1991  1992  1993  1994  1995  1996  1997  \
Bhutan       0     0     1     0     2     2     1     1     4     2     2   
Maldives     0     0     0     0     0     3     3     0     0     0     0   

          1998  1999  2000  2001  2002  2003  2004  2005  2006  2007  2008  \
Bhutan       1     3     6     6     8     7     1     5    10     7    36   
Maldives     1     0     1     0     1     0     1     0     0     2     1   

          2009  2010  2011  2012  2013  Total  
Bhutan     865  1464  1879  1075   487   5876  
Maldives     7     4     3     1     1     30

<!--NAVIGATION-->
< [Previous](https://github.com/Egade/ClassNotes/blob/main/021_read_iloc_unique_groupby.ipynb) 
| [Toc](https://github.com/Egade/ClassNotes) 
| [Next](nex) >